[WIP] Add UPDATE sql support in Presto Engine

agrawalreetika opened the pull request:
## Description
Add UPDATE SQL support in Presto Engine.
PR also includes new developer documentation which has details about how it works

## Motivation and Context
#20571

## Impact
- New Support

## Test Plan
- Tested Parser changes and added tests for the same
- Once we add support for UPDATE for any connector, then complete flow could be tested via Product test added in [AbstractTestDistributedQueries.java](https://github.com/prestodb/presto/compare/master...agrawalreetika:update-support?expand=1#diff-e2539b884864825c29df8d11f0d91b146d4086303736b32addd9cdd44a36fee1)

## Contributor checklist

- [ ] Please make sure your submission complies with our [development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), [commit message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests), and [attribution guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution).
- [ ] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [ ] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [ ] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [ ] Adequate tests were added if applicable.
- [ ] CI passed.

## Release Notes

```
== RELEASE NOTES ==

General Changes
* Add UPDATE sql support in Presto Engine
```

steveburnett commented on presto-docs/src/main/sphinx/develop/delete-and-update.rst:
```diff
@@ -0,0 +1,178 @@
+
+* For each split, the connector will create an ``UpdatablePageSource`` instance, layered over the
+  connector's ``ConnectorPageSource``, to read pages on behalf of the Presto engine, and to
+  write deletions and/or updates to the underlying data store.
```
```suggestion
  write deletions or updates to the underlying data store.
```
From the [GitLab recommended word list entry for and/or](https://docs.gitlab.com/ee/development/documentation/styleguide/word_list.html#andor), 

"Instead of and/or, use or or rewrite the sentence to spell out both options."

steveburnett commented on presto-docs/src/main/sphinx/develop/delete-and-update.rst:
```diff
@@ -0,0 +1,178 @@
+  ``finishUpdate``, passing a collection containing all the "fragments" from all the splits.  The connector
+  does what is required to finalize the operation, for example, committing the transaction.
+
+The rowId Column Abstraction
```
Consider adding "handle" to the heading titles in lines 34, 41, and 52, to help the reader find the explanation of this concept.

steveburnett commented on presto-docs/src/main/sphinx/develop/delete-and-update.rst:
```diff
@@ -0,0 +1,178 @@
+The rowId Column Abstraction
+============================
+
+The Presto engine and connectors use a "rowId" column handle abstraction to agree on the identities of rows
```
As mentioned above, suggest no quote marks around "rowId".

steveburnett commented on presto-docs/src/main/sphinx/develop/delete-and-update.rst:
```diff
@@ -0,0 +1,178 @@
+=======================
+
+As mentioned above, to support ``DELETE`` or ``UPDATE``, the connector must define a subclass of
+``UpdatablePageSource``, layered over the connector's usual ``ConnectorPageSource``.  The interesting methods are:
```
```suggestion
``UpdatablePageSource``, layered over the connector's ``ConnectorPageSource``.  The interesting methods are:
```
As above, suggest deleting "usual" - unless there is a meaning to its presence that I do not understand. If "usual" is meaningful, can you help me to understand why?

steveburnett reviewed (changes requested):
This is great doc, thanks! 

I made some rephrasing and formatting suggestions, some of which are only suggestions for you to consider if you think they help. As always, if I've misunderstood your meaning, help me understand and we will find a solution.

(Oh, there are no comments about sql/update.rst because I didn't have anything to say about that file. I'm just mentioning it here to say that I didn't miss it :) ).

yingsu00 commented on presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java:
```diff
@@ -289,6 +296,105 @@ public DeleteNode plan(Delete node)
+    public UpdateNode plan(Update node)
+    {
+        RelationType descriptor = analysis.getOutputDescriptor(node.getTable());
+        Table table = node.getTable();
```
Just inline table

yingsu00 commented on presto-spi/src/main/java/com/facebook/presto/spi/plan/UpdateNode.java:
```diff
@@ -0,0 +1,114 @@
+    private final VariableReferenceExpression rowId;
+    private final List<VariableReferenceExpression> columnValueAndRowIdSymbols;
+    private final List<VariableReferenceExpression> outputVariables;
+    @JsonCreator
```
nit: add empty line above

yingsu00 reviewed (commented):
@agrawaldevesh Was this ported from Trino? If yes can you please add link to the Trino commit/PR?

steveburnett reviewed (approved):
LGTM! (docs)

Thanks!

agrawalreetika commented:
@yingsu00, I had added Trino commit in the commit message, I have now added it to PR description as well.
I have fixed the tests as well, please review.

agrawalreetika commented on presto-spi/pom.xml:
```diff
@@ -88,7 +88,7 @@
             <groupId>com.google.guava</groupId>
             <artifactId>guava</artifactId>
-            <scope>test</scope>
+            <scope>provided</scope>
```
It is to resolve imports in [UpdateNode](https://github.com/prestodb/presto/pull/21435/files#diff-e771f381c13140371738c2c2ffc7a1af28c5f3022a773a477aaa19e91dacd14b) Class

agrawalreetika replied:
Yes @yingsu00, I noticed this refactor was done in Presto. So I have also moved UpdateNode as well in same `presto-main` module itself.

yingsu00 commented:
Reetika, UpdatablePageSource was removed in https://github.com/trinodb/trino/pull/15161/. Will you be able to check if it's applicable to Presto? Iceberg does not use this page source.
